llpc/util/llpcElfWriter.cpp

```diff
@@ -1118,6 +1206,155 @@ void ElfWriter<Elf>::Reinitialize()
+        break;
+    }
+
+    if (isGraphicsPipeline)
+    {
+        auto pPipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo*>(pContext->GetPipelineBuildInfo());
+        *pValue = GetDescriptorResourceOffset(descSet, binding, type, pPipelineInfo->vs.pUserDataNodes,
+                                              pPipelineInfo->vs.userDataNodeCount);
+        if (*pValue == InvalidValue)
+        {
+            *pValue = GetDescriptorResourceOffset(descSet, binding, type, pPipelineInfo->fs.pUserDataNodes,
+                                                  pPipelineInfo->fs.userDataNodeCount);
+        }
+    }
+    else
+    {
+        auto pPipelineInfo = reinterpret_cast<const ComputePipelineBuildInfo*>(pContext->GetPipelineBuildInfo());
+        *pValue = GetDescriptorResourceOffset(descSet, binding, type, pPipelineInfo->cs.pUserDataNodes,
+                                              pPipelineInfo->cs.userDataNodeCount);
+    }
+
+    return *pValue != InvalidValue;
+}
+
+// =====================================================================================================================
+// Get the value of a relocation symbol. Returns true if success, and false if the symbol is unknown.
+bool GetRelocationSymbolValue(
+    Context* pContext,       // [in] Pipeline compilation context
+    RelocationEntry reloc,   // The relocation entry
+    bool isGraphicsPipeline, // Whether we are processing a graphics pipeline
+    uint32_t* pValue)         // [out] the value of the relocation symbol
+{
+    assert(reloc.type == R_AMDGPU_ABS32 && "can only handle R_AMDGPU_ABS32 typed relocations.");
+    if (strncmp(reloc.name, "doff_", 5) == 0)
+    {
+        return GetDescriptorOffsetRelocationValue(pContext, reloc, isGraphicsPipeline, pValue);
+    }
+    return false;
+}
+
+// =====================================================================================================================
+// Fixup relocations in the ELF with actual values.
+template<class Elf>
+void FixUpRelocations(
+    ElfWriter<Elf>* pWriter,                               // [in,out] Writer to target ELF
+    const std::vector<RelocationEntry>& relocations,       // [in] Relocation entries
+    Context* pContext,                                     // [in] Pipeline compilation context
+    bool isGraphicsPipeline)                               // Whether we are fixing a compute or graphics pipeline
+{
+    char* pData = nullptr;
+    size_t dataLength = 0;
+    pWriter->GetSectionData(TextName, const_cast<const void**>(reinterpret_cast<void **>(&pData)), &dataLength);
+
+    for(uint32_t i = 0; i < relocations.size(); ++i)
+    {
+        auto reloc = relocations[i];
+        uint32_t relocationValue = 0;
+        if (GetRelocationSymbolValue(pContext, reloc, isGraphicsPipeline, &relocationValue))
+        {
+            assert(pData != nullptr && dataLength >= reloc.offset);
+            memcpy(pData + reloc.offset, &relocationValue, 4);
+        }
```

Comment:
Could you please make this assert when the symbol value is not found and when the relocation type is not as expected?

Also, this doesn't yet take the addend into account, does it?

Reply:
Addends are being handled in `ElfReader<>::GetRelocation()`. Since currently all relocations are read from ".rel" section, there is no addends included. If in the future we are going to have "rela" section support, the addends can be handled there.

Reply:
This is not correct: When a .rel section is used, the addend is taken from the target section.

So for example, if the symbol value is 160 and the target dword in the target section is not 0 but instead e.g. 32, then the resulting value in the target section should be 192 for most relocation types.

Reply:
Where is this "target dword" defined?

Reply:
Here's the `Reloc` structure defined in `llpcElfReader.h`:
```
   // ELF relocation entry
    struct Reloc
    {
        uint64_t r_offset;     // Location (file byte offset, or program virtual address)
        union
        {
            uint64_t r_info;   // Symbol table index and type of relocation to apply
            struct
            {
                uint32_t  r_type;   // Type of relocation
                uint32_t  r_symbol; // Index of the symbol in the symbol table
            };
        };
    };
```

I cannot find anything related to "target" here, or in `SectionHeader` or in `Symbol`.

Reply:
Do you refer to the dword value itself at the `offset` location? i.e. the true offset is
`actualOffset = *(uint32_t*)(pData+rel.offset) + rel.offset` ?

Reply:
This dropped off my radar, but yes, that's what I meant.